### PR TITLE
HIVE-25102: Cache Iceberg table objects within same query

### PR DESCRIPTION
**Pull request by lcspinter**
### What changes were proposed in this pull request?
Introduce a caching layer to store iceberg table objects used within the same query. What is covered in this PR:
- Enhance the SessionState class with the ability to store QueryState objects bound to the query ID.
- Add caching layer to the QueryState class. The cache keys are table identifiers (database name + table name) and the values are iceberg table objects
- Store QueryState instance in Driver class, when the query execution is started
- Create utility class, which handles the loading of iceberg tables either from the cache or from the catalog through the Catalogs API
- Change Catalogs.loadTable() calls to IcebergTableUtil.getTable()
- Clean up resources (table objects from QueryState and QueryState references from SessionState) when query execution is finished

### Why are the changes needed?
Catalogs.loadTable(configuration, props) is called numerous times within the same query which can be expensive. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manual test

### CODE REVIEW

**Comment by lcspinter** [comment]
@marton-bod @pvary @szlta Could you please review this PR? Thanks

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 1]
```diff
@@ -284,4 +333,40 @@ private static String getCatalogType(Configuration conf, String catalogName) {
       }
     }
   }
+
+  public static class TableCache {
+    private static final Cache<String, Table> tableCache = Caffeine.newBuilder()
+        .expireAfterAccess(12, TimeUnit.HOURS).build();
+    private static final Set<String> keys = new TreeSet<>();
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:340]: Instead of maintaining the key set manually, you might be able to get the keys in the hook method by using `tableCache.getNativeCache().asMap().keySet()`?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 2]
```diff
@@ -284,4 +333,40 @@ private static String getCatalogType(Configuration conf, String catalogName) {
...
+
+    public static void removeTable(Configuration conf) {
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:342]: Can we rename this to make it clear that it removes _all_ tables for a given query? `removeTables`, `removeTablesForQuery`, or something similar that you'd prefer.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 3]
```diff
@@ -284,4 +333,40 @@ private static String getCatalogType(Configuration conf, String catalogName) {
...
+      if (queryId != null && !queryId.isEmpty()) {
+        Set<String> queryKeys = keys.stream().filter(k -> k.startsWith(queryId)).collect(Collectors.toSet());
+        tableCache.invalidateAll(queryKeys);
+        keys.removeAll(queryKeys);
+      }
+    }
+
+    public static Table getTable(Configuration conf, String catalogName, String tableIdentifier) {
+      String queryId = conf.get(HiveConf.ConfVars.HIVEQUERYID.varname);
+      if (queryId != null  && !queryId.isEmpty()) {
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:353]: nit: double spaces after the queryId != null part, here and below

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergQueryLifeTimeHook.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+public class HiveIcebergQueryLifeTimeHook implements QueryLifeTimeHook {
+  @Override
+  public void beforeCompile(QueryLifeTimeHookContext ctx) {
+
+  }
+
+  @Override
+  public void afterCompile(QueryLifeTimeHookContext ctx, boolean hasError) {
+  }
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergQueryLifeTimeHook.java:34]: nit: missing new line, like for beforeCompile and afterExecution

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 5]
```diff
@@ -284,4 +333,40 @@ private static String getCatalogType(Configuration conf, String catalogName) {
...
+        .expireAfterAccess(12, TimeUnit.HOURS).build();
+    private static final Set<String> keys = new TreeSet<>();
+
+    public static void removeTable(Configuration conf) {
+      String queryId = conf.get(HiveConf.ConfVars.HIVEQUERYID.varname);
+      if (queryId != null && !queryId.isEmpty()) {
+        Set<String> queryKeys = keys.stream().filter(k -> k.startsWith(queryId)).collect(Collectors.toSet());
+        tableCache.invalidateAll(queryKeys);
+        keys.removeAll(queryKeys);
+      }
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:348]: shall we add an else branch here with some warn logging? It might be a problem here worth noting if the query id is not present in the config, because that will cause cache pollution over time. In the other methods, if the query id is not present we just reload the table, so that's not as big of a deal

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergQueryLifeTimeHook.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+  @Override
+  public void beforeExecution(QueryLifeTimeHookContext ctx) {
+    Catalogs.TableCache.removeTable(ctx.getHiveConf());
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergQueryLifeTimeHook.java:38]: Shouldn't we call this in `afterCompile`? If there's a compilation error at some point, we won't even get to this method, no?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 7]
```diff
@@ -284,4 +333,40 @@ private static String getCatalogType(Configuration conf, String catalogName) {
       }
     }
   }
+
+  public static class TableCache {
+    private static final Cache<String, Table> tableCache = Caffeine.newBuilder()
+        .expireAfterAccess(12, TimeUnit.HOURS).build();
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:339]: Shall we decrease this a bit? Perhaps 30 minutes would also suffice? Not sure what a "long" compilation time means for Hive usually. Decreasing it could give us some extra safety against cache pollution in case the hook-based eviction doesn't always go smoothly. Worst case in a very long compilation scenario, we might reload the table again eventually, but I'm assuming compilations that take 30+ minutes should be pretty rare?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 1 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:340]: Right, I don't know how I missed that. I was looking for a map representation of the cache entries.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 5 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:348]: Correct, added logging.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergQueryLifeTimeHook.java (hunk 6 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergQueryLifeTimeHook.java:38]: I added it to `afterCompile` and `afterExecution` as well.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 7 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:339]: I kept the 12 hours threshold since we want to keep the cache entries until the execution is finished.

**Review by marton-bod — APPROVED** [review]

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 8]
```diff
@@ -284,4 +334,42 @@ private static String getCatalogType(Configuration conf, String catalogName) {
...
+    private static final Logger LOG = LoggerFactory.getLogger(TableCache.class);
+
+    public static void removeTables(Configuration conf) {
+      String queryId = conf.get(HiveConf.ConfVars.HIVEQUERYID.varname);
+      if (queryId != null && !queryId.isEmpty()) {
+        Set<String> queryKeys = tableCache.asMap().keySet().stream()
+            .filter(k -> k.startsWith(queryId)).collect(Collectors.toSet());
+        tableCache.invalidateAll(queryKeys);
+      } else {
+        LOG.warn("Query id is not present in config, therefore no Iceberg table object is removed " +
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:350]: Is this really a `warn`? Do we expect at least one table it the cache every time?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 9]
```diff
@@ -284,4 +334,42 @@ private static String getCatalogType(Configuration conf, String catalogName) {
...
+
+    public static Table getTable(Configuration conf, String catalogName, String tableIdentifier) {
+      String queryId = conf.get(HiveConf.ConfVars.HIVEQUERYID.varname);
+      if (queryId != null  && !queryId.isEmpty()) {
+        return tableCache.getIfPresent(getKey(queryId, catalogName, tableIdentifier));
+      }
+      return null;
+    }
+
+    public static void addTable(Configuration conf, String catalogName, String tableIdentifier, Table table) {
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:363]: Why not use lambda to generate the table object if not present? Like it is done with `putIfAbsent`?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 10]
```diff
@@ -284,4 +334,42 @@ private static String getCatalogType(Configuration conf, String catalogName) {
       }
     }
   }
+
+  public static class TableCache {
+    private static final Cache<String, Table> tableCache = Caffeine.newBuilder()
+        .expireAfterAccess(12, TimeUnit.HOURS).build();
+    private static final Logger LOG = LoggerFactory.getLogger(TableCache.class);
+
+    public static void removeTables(Configuration conf) {
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:343]: Do we need the whole `conf` here? Shouldn't we separate out the concerns and use only an `executionId` instead?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 11]
```diff
@@ -284,4 +334,42 @@ private static String getCatalogType(Configuration conf, String catalogName) {
...
+            .filter(k -> k.startsWith(queryId)).collect(Collectors.toSet());
+        tableCache.invalidateAll(queryKeys);
+      } else {
+        LOG.warn("Query id is not present in config, therefore no Iceberg table object is removed " +
+            "from the table cache.");
+      }
+    }
+
+    public static Table getTable(Configuration conf, String catalogName, String tableIdentifier) {
+      String queryId = conf.get(HiveConf.ConfVars.HIVEQUERYID.varname);
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:356]: Do we need the whole `conf` here? Shouldn't we separate out the concerns and use only an `executionId` instead?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java [hunk 12]
```diff
@@ -307,7 +307,7 @@ private void commitTable(FileIO io, ExecutorService executor, JobContext jobCont
     if (catalogName != null) {
       catalogProperties.put(InputFormatConfig.CATALOG_NAME, catalogName);
     }
-    Table table = Catalogs.loadTable(conf, catalogProperties);
+    Table table = Catalogs.loadTableSkipCache(conf, catalogProperties);
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:310]: Should it be `Catalogs.invalidate` if the catalog implements `CachingCatalog` interface?
This way we break the compatibility with other Catalogs.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 8 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:350]: No, we don't except, but I guess we should signal that the query id is not present config, which could lead to cache pollution.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 9 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:363]: The `tableCache` doesn't have a put method that accepts functional interfaces.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 10 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:343]: Yes, it makes sense.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java (hunk 12 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:310]: I'm not sure I understand your point here. Could you please add a bit more details?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 9 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:363]: I was trying to suggest to add such a method to the `TableCache` class. I think it could make the code easier to read. What do you think?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 8 above)
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:350]: I think what we do expect is that the `hive.query.id` is always populated in the config. If it's absent, that's a warning sign and something which could prevent us from evicting from the cache. This could be a useful log line in case we see that the cache gets overly large and we're looking for the root cause why the iceberg tables are not evicted.

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java [hunk 13]
```diff
@@ -349,6 +349,20 @@
 
   private final AtomicLong sparkSessionId = new AtomicLong();
 
+  private final Map<String, Object> queryStateMap = new HashMap<>();
+
+  public Object getQueryState(String queryId) {
+    return queryStateMap.get(queryId);
+  }
+
+  public void addQueryState(String queryId, Object queryState) {
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:358]: Shouldn't we only accept only QueryState objects here?
i.e. `public void addQueryState(String queryId, QueryState queryState) {`

File: ql/src/java/org/apache/hadoop/hive/ql/QueryState.java [hunk 14]
```diff
@@ -59,6 +60,11 @@
 
   static public final String USERID_TAG = "userid";
 
+  /**
+   * map of tables involved in the query.
+   */
+  private final Map<String, Object> tableMap = new HashMap<>();
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/QueryState.java:66]: Can we make this a generic map so we can store things other than tables here? We will need this container also for other info that - in the absence of this new query state feature - we previously had to force into the conf, such as write commit info (jobID, vertexID, taskNum), CTAS info (is the query ctas? what's the ctas target table name), etc.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 15]
```diff
@@ -97,13 +98,15 @@ public static Table loadTable(Configuration conf, Properties props) {
             props.getProperty(InputFormatConfig.CATALOG_NAME));
   }
 
-  private static Table loadTable(Configuration conf, String tableIdentifier, String tableLocation,
-                                 String catalogName) {
+  private static Table loadTable(Configuration conf, String tableIdentifier, String tableLocation, String catalogName) {
     Optional<Catalog> catalog = loadCatalog(conf, catalogName);
 
+    Table cachedTable = null;
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:104]: Are these changes still necessary?

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java (hunk 13 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java]: Right. Changed it.

File: ql/src/java/org/apache/hadoop/hive/ql/QueryState.java (hunk 14 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/QueryState.java]: Changed the naming of the map.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 15 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java]: Not at all. The revert was not 100% successful.

File: ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java [hunk 16]
```diff
@@ -349,6 +350,20 @@
 
   private final AtomicLong sparkSessionId = new AtomicLong();
 
+  private final Map<String, QueryState> queryStateMap = new HashMap<>();
+
+  public Object getQueryState(String queryId) {
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java:355]: Can we update the return type too so we don't need any casting?

Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java]: Sure!

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+   * @param properties controlling properties
+   * @return
+   */
+  static Table getTable(Configuration configuration, Properties properties) {
+    // look for the table object stored in the query state. If it's null, it means the table was not loaded yet
+    // within the same query therefore we claim it through the Catalogs API and then store it in query state.
+    QueryState queryState = SessionState.get()
+        .getQueryState(configuration.get(HiveConf.ConfVars.HIVEQUERYID.varname));
+    Table table = null;
+    if (queryState != null) {
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:48]: Do we expect the query state to be null at any point during compilation? If so, that would result in constant reloading of the table. So maybe we should either remove the null checks to simplify the code, or at least log it if it's not there? What do you think?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  /**
+   * Load the iceberg table either from the {@link QueryState} or through the configured catalog.
+   * @param configuration a Hadoop configuration
+   * @param properties controlling properties
+   * @return
+   */
+  static Table getTable(Configuration configuration, Properties properties) {
+    // look for the table object stored in the query state. If it's null, it means the table was not loaded yet
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:43]: nit: I think this comment belongs to the javadoc more naturally, but it's up to you, I don't feel strongly on that

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java (hunk 17 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:48]: Right now we don't expect it to be null, but I'm not sure about the availability of SessionState object. I noticed this comment in the class:  `SessionState is not available in runtime and Hive.get().getConf() is not safe to call`.
So I put there this safeguard, to be more robust in case the SessionState is lost, but I agree some additional logging doesn't hurt.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 19]
```diff
@@ -97,13 +97,13 @@ public static Table loadTable(Configuration conf, Properties props) {
...
 
-  private static Table loadTable(Configuration conf, String tableIdentifier, String tableLocation,
-                                 String catalogName) {
+  private static Table loadTable(Configuration conf, String tableIdentifier, String tableLocation, String catalogName) {
     Optional<Catalog> catalog = loadCatalog(conf, catalogName);
 
     if (catalog.isPresent()) {
       Preconditions.checkArgument(tableIdentifier != null, "Table identifier not set");
-      return catalog.get().loadTable(TableIdentifier.parse(tableIdentifier));
+      Table table = catalog.get().loadTable(TableIdentifier.parse(tableIdentifier));
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:105]: Is this change needed?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 20]
```diff
@@ -87,8 +87,8 @@ public static Table loadTable(Configuration conf) {
    * The table identifier ({@link Catalogs#NAME}) and the catalog name ({@link InputFormatConfig#CATALOG_NAME}),
    * or table path ({@link Catalogs#LOCATION}) should be specified by the controlling properties.
    * <p>
-   * Used by HiveIcebergSerDe and HiveIcebergStorageHandler
-   * @param conf a Hadoop
+   * Used by HiveIcebergSerDe and HiveIcebergStorageHandler.
+   * @param conf a Hadoop configuration
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:91]: MAybe we can revert all of the changes in this file?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+  /**
+   * Load the iceberg table either from the {@link QueryState} or through the configured catalog. Look for the table
+   * object stored in the query state. If it's null, it means the table was not loaded yet within the same query
+   * therefore we claim it through the Catalogs API and then store it in query state.
+   * @param configuration a Hadoop configuration
+   * @param properties controlling properties
+   * @return
+   */
+  static Table getTable(Configuration configuration, Properties properties) {
+    QueryState queryState = SessionState.get()
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:49]: Is it possible for `SessionState.get()` to return null? What happens on TezAM side?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+        .getQueryState(configuration.get(HiveConf.ConfVars.HIVEQUERYID.varname));
+    Table table = null;
+    String tableIdentifier = properties.getProperty(Catalogs.NAME);
+    if (queryState != null) {
+      table = (Table) queryState.getResource(tableIdentifier);
+    } else {
+      LOG.debug("QueryState is not available in SessionState. Loading {} from configured catalog.", tableIdentifier);
+    }
+    if (table == null) {
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:58]: nit: new lines

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+      table = Catalogs.loadTable(configuration, properties);
+      if (queryState != null) {
+        queryState.addResource(tableIdentifier, table);
+      }
+    }
+    return table;
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:64]: nit: new line

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java (hunk 17 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:53]: Maybe Optional<QueryState> ?

Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:53]: In this case, wrapping the QueryState in an Optional doesn't add much value.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 19 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:105]: Reverted the whole file.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java (hunk 21 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:49]: I added an additional check for null SessionState. This method is called only from HS2 side.

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java [hunk 24]
```diff
@@ -213,7 +216,14 @@ private void runInternal(String command, boolean alreadyCompiled) throws Command
         releaseResources();
       }
 
+      if (SessionState.get() != null) {
+        // Clean up every resource object stored in the query state
+        driverContext.getQueryState().removeResources();
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/Driver.java:221]: Do we need this one?
We remove the `QueryState` anyway...
And then we do not need `removeResources` either

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 20 above)
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:91]: MAybe revert this one too?

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java (hunk 24 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java:221]: I think we need it. We have an instance of `QueryState` stored on the `DriverContext`, and we have the same `QueryState` instance stored on `SessionState` as well. If I remove the `QueryState` just from the `SessionState`, we will end up with some dangling resources on the `QueryState` bound to the `DriverContext`.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 20 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/Catalogs.java:91]: I would rather keep it, since it corrects the Javadoc.

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java (hunk 24 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java:221]: The QueryState in the DriverContext should be removed as well when the query finishes. Otherwise we still would have a leaked QueryState which is a memory leak itself. So this only makes the leak smaller, but does not fix it (if we have any)

Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java]: I've deleted the `removeResources`.

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java [hunk 25]
```diff
@@ -213,7 +216,12 @@ private void runInternal(String command, boolean alreadyCompiled) throws Command
         releaseResources();
       }
 
+      if (SessionState.get() != null) {
+        // Remove any query state reference from the session state
+        SessionState.get().removeQueryState(getConf().get(HiveConf.ConfVars.HIVEQUERYID.varname));
+      }
       driverState.executionFinishedWithLocking(isFinishedWithError);
+
     }
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/Driver.java:225]: nit of the nit for the newlines:
```suggestion
      }

      if (SessionState.get() != null) {
        // Remove any query state reference from the session state
        SessionState.get().removeQueryState(getConf().get(HiveConf.ConfVars.HIVEQUERYID.varname));
      }

      driverState.executionFinishedWithLocking(isFinishedWithError);
    }
```

**Review by pvary — APPROVED** [review]

**Comment by marton-bod** [comment]
@lcspinter Can you please replace the Catalogs.loadTable call in the `HiveIcebergStorageHandler#overlayTableProperties` as well?
